Updating the O3DE Memory Management documentation (#2503)
o3de/o3de.org: content/docs/user-guide/programming/memory-management.md

+ Specify the allocator in each class.
+ Use child allocators. To tag and track resource usage, new gems and subsystems should create their own allocator or create a `ChildAllocatorSchema` that references an existing allocator. For an example of creating a child allocator, see [Creating an Allocator](/docs/user-guide/programming/memory/allocators/).

**Reason**: O3DE's core AZ systems provide a memory managed environment, not a raw system allocator like the managed memory in C\# or Java. In O3DE, the core AZ systems provide speed, safety, and facilities for tracking memory usage.

For information about O3DE's scheme for allocators and new allocators, see [Manually Allocating Memory](/docs/user-guide/programming/memory/allocators/).

## O3DE Memory System Initialization

O3DE initializes its memory system via storing an instances of Allocators in an AZ Environment variables the first time a call to `AZ::AllocatorInstance<AllocatorType>::Get()` is called in [code](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Memory/AllocatorInstance.h#L73-L86).  

> To ensure that Allocators are available in static initialization and static de-initialization of the running applicaition and any dynamically loaded Gem modules, the allocator AZ Environment variables are stored in the `O3DEKernel` shared library which link library dependency of the [AzCore Target](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/CMakeLists.txt#L60-L98).  

Reviewer: ```suggestion
To ensure that Allocators are available in static initialization and static de-initialization of the running applicaition and any dynamically loaded Gem modules, the allocator AZ Environment variables are stored in the `O3DEKernel` shared library which is a link library dependency of the [AzCore Target](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/CMakeLists.txt#L60-L98).  
```

> To ensure that Allocators are available in static initialization and static de-initialization of the running applicaition and any dynamically loaded Gem modules, the allocator AZ Environment variables are stored in the `O3DEKernel` shared library which is a link library dependency of the [AzCore Target](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/CMakeLists.txt#L60-L98).  

Reviewer 2: ```suggestion
To ensure that Allocators are available in static initialization and static de-initialization of the running application and any dynamically loaded Gem modules, the allocator AZ Environment variables are stored in the `O3DEKernel` shared library which is a link library dependency of the [AzCore Target](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/CMakeLists.txt#L60-L98).  
```

> As the `AzCore` static library target is a dependency by Gems and O3DE applications, they all statically link the `O3DEKernel` target due to that dependency.  

Reviewer: Possible rewording?
```suggestion
All Gems and O3DE applications statically link the `O3DEKernal` target because they  depend on the `AzCore` static library target.
```

Reviewer 3: The re-write sounds good.

> On initial load of the application, the `O3DEKernel` library is the first shared library loaded.  

Reviewer 2: `load` is redundant.

```suggestion
On initial startup of the application, the `O3DEKernel` library is the first shared library loaded.  
```

> As loading of statically linked shared libraries occur before static initialization and unloading occurs after static de-initialization of the application, this ensures that the O3DE Allocators are available for the entire duration of the application.  

Reviewer: possible rewording - up to you which you think reads better
```suggestion
O3DE Allocators are available for the entire duration of the application because loading of statically linked shared libraries occur before static initialization and unloading occurs after static de-initialization of the application.
```

Reviewer 3: Your re-write is in an active voice, which is inline with the documentation standards.

> The [AZ::Environnment](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Module/Environment.h#L114-L115) is accessed through the O3DEKernel via a DLL interface to enforce the only instance to be available in `O3DEKernel.dll` (Windows), `libO3DEKernel.so` (Linux/Android) or `libO3DEKernel.dylib` (MacOS/iOS).  

Reviewer: As translated from ancient O3DE texts:
```
One AZ::Environment API to rule them all, 
One AZ::Environment API to find them, 
One AZ::Environment API to bring them all, 
and in the O3DEKernal bind them.
```

```suggestion
Only one [AZ::Environnment](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Module/Environment.h#L114-L115) instance exists and is accessed through the O3DEKernel via a DLL interface in `O3DEKernel.dll` (Windows), `libO3DEKernel.so` (Linux/Android) or `libO3DEKernel.dylib` (MacOS/iOS) to enforce the use of the one `AZ::Environment` (see translation of ancient inscription).
```

Reviewer 3: We should leave the ancient inscription part out of it.

Reviewer 3: ```suggestion
Only one [AZ::Environnment](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Module/Environment.h#L114-L115) instance exists and is accessed through the O3DEKernel via a DLL interface in `O3DEKernel.dll` (Windows), `libO3DEKernel.so` (Linux/Android) or `libO3DEKernel.dylib` (MacOS/iOS) to enforce the use of the one `AZ::Environment`.```

Reviewer: If you don't want this documentation on the New York Times best seller list...
